Win32-network/src/System/Win32/Async/IOData.hsc

```diff
@@ -0,0 +1,155 @@
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+#include <windows.h>
+#include <Win32-network.h>
+
+module System.Win32.Async.IOData
+  ( IOData (..)
+  , iodOverlappedPtr
+  , iodDataPtr
+
+  , Result (..)
+  , withIODataPtr
+  ) where
+
+import Control.Concurrent ( MVar
+                          , newEmptyMVar
+                          )
+import Control.Exception (mask)
+import Control.Monad (when)
+import Foreign ( Ptr
+               , StablePtr
+               , Storable (..)
+               , castPtr
+               , newStablePtr
+               , nullPtr
+               , plusPtr
+               )
+
+import           System.Win32.Types (ErrCode)
+import qualified System.Win32.Types as Win32
+import qualified System.Win32.Mem   as Win32
+
+import System.Win32.Async.Overlapped
+
+
+-- | Data associated with each async call.  This is represented as the 'IODATA'
+-- struct.  The windows kernel macro 'CONTAINING_RECORD' allows to access the
+-- 'IODATA' when the io action is signaled 'GetQueuedCompletionStatus' via the
+-- 'iodOverlapped :: OVERLAPPED' member.
```

Comment:
If we decide not to use that macro and do it Haskell side, we can update this comment.